File: app/src/main/java/openfoodfacts/github/scrachx/openfood/features/searchbycode/SearchByCodeFragment.java
package openfoodfacts.github.scrachx.openfood.features.searchbycode;

import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;

import org.apache.commons.lang.StringUtils;

import openfoodfacts.github.scrachx.openfood.R;
import openfoodfacts.github.scrachx.openfood.databinding.FragmentFindProductBinding;
import openfoodfacts.github.scrachx.openfood.features.shared.NavigationBaseFragment;
import openfoodfacts.github.scrachx.openfood.network.OpenFoodAPIClient;
import openfoodfacts.github.scrachx.openfood.utils.NavigationDrawerListener.NavigationDrawerType;
import openfoodfacts.github.scrachx.openfood.utils.ProductUtils;
import openfoodfacts.github.scrachx.openfood.utils.Utils;

import static openfoodfacts.github.scrachx.openfood.utils.NavigationDrawerListener.ITEM_SEARCH_BY_CODE;

/**
 * @see R.layout#fragment_find_product
 */
public class SearchByCodeFragment extends NavigationBaseFragment {
    public static final String INTENT_KEY_BARCODE = "barcode";
    private FragmentFindProductBinding binding;
    private OpenFoodAPIClient api;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        api = new OpenFoodAPIClient(requireActivity());
        binding = FragmentFindProductBinding.inflate(inflater);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        binding.editTextBarcode.setSelected(false);
        binding.buttonBarcode.setOnClickListener(v -> checkBarcodeThenSearch());

        Intent intent = requireActivity().getIntent();
        if (intent != null) {
            String barCode = intent.getStringExtra(INTENT_KEY_BARCODE);
            if (StringUtils.isNotEmpty(barCode)) {
                setBarcodeThenSearch(barCode);
            }
        }
    }

    private void setBarcodeThenSearch(String code) {
        binding.editTextBarcode.setText(code, TextView.BufferType.EDITABLE);
        checkBarcodeThenSearch();
    }

    private void checkBarcodeThenSearch() {
        Utils.hideKeyboard(requireActivity());

        final String barCodeTxt = binding.editTextBarcode.getText().toString();
        if (barCodeTxt.isEmpty()) {
            binding.editTextBarcode.setError(getResources().getString(R.string.txtBarcodeRequire));
        } else if (!ProductUtils.isBarcodeValid(barCodeTxt)) {
            binding.editTextBarcode.setError(getResources().getString(R.string.txtBarcodeNotValid));
        } else {
            api.openProduct(barCodeTxt, getActivity());
        }
    }

    @Override
    @NavigationDrawerType
    public int getNavigationDrawerType() {
        return ITEM_SEARCH_BY_CODE;
    }

    @Override
    public void onResume() {
        super.onResume();
        final ActionBar supportActionBar = ((AppCompatActivity) requireActivity()).getSupportActionBar();
        if (supportActionBar != null) {
            supportActionBar.setTitle(getString(R.string.search_by_barcode_drawer));
        }
    }
}

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/features/searchbycode/SearchByCodeFragment.kt
package openfoodfacts.github.scrachx.openfood.features.searchbycode

import android.os.Bundle
import android.view.KeyEvent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.EditorInfo
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.viewModels
import dagger.hilt.android.AndroidEntryPoint
import openfoodfacts.github.scrachx.openfood.R
import openfoodfacts.github.scrachx.openfood.databinding.FragmentFindProductBinding
import openfoodfacts.github.scrachx.openfood.features.shared.NavigationWithDrawerBaseFragment
import openfoodfacts.github.scrachx.openfood.utils.*
import openfoodfacts.github.scrachx.openfood.utils.NavigationDrawerListener.NavigationDrawerType

/**
 * @see R.layout.fragment_find_product
 */
@AndroidEntryPoint
class SearchByCodeFragment : NavigationWithDrawerBaseFragment() {
    private var _binding: FragmentFindProductBinding? = null
    private val binding get() = _binding!!

    private val viewModel: SearchByCodeViewModel by viewModels()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        _binding = FragmentFindProductBinding.inflate(inflater)
        return binding.root
    }

    override fun onDestroyView() {
        removeDrawerListener()
        super.onDestroyView()
        _binding = null
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.editTextBarcode.let { e ->
            e.isSelected = false

            // Accept both the Enter key (from the keyboard) or the search virtual key
            e.setOnEditorActionListener { _, actionId, keyEvent ->
                if (actionId == EditorInfo.IME_ACTION_SEARCH ||
                    keyEvent.keyCode == KeyEvent.KEYCODE_ENTER
                ) {
                    checkBarcodeThenSearch()
                    true
                } else {
                    false
                }
            }
        }


        binding.buttonBarcode.setOnClickListener { checkBarcodeThenSearch() }

        // Get barcode from intent or saved instance or from arguments, in this order
        val barcode = requireActivity().intent.getStringExtra(INTENT_KEY_BARCODE).takeUnless { it.isNullOrBlank() }
            ?: savedInstanceState?.getString(INTENT_KEY_BARCODE).takeUnless { it.isNullOrBlank() }
            ?: arguments?.getString(INTENT_KEY_BARCODE).takeUnless { it.isNullOrBlank() }

        // If we have the barcode, set the textbox and start the search
        if (!barcode.isNullOrBlank()) {
            binding.editTextBarcode.setText(barcode, TextView.BufferType.EDITABLE)
            checkBarcodeThenSearch()
        }
    }

    private fun checkBarcodeThenSearch() {
        requireActivity().hideKeyboard()

        val code = binding.editTextBarcode.text.toString()
        when {
            code.isEmpty() -> {
                binding.editTextBarcode.error = resources.getString(R.string.txtBarcodeRequire)
            }
            !isBarcodeValid(code) -> {
                binding.editTextBarcode.error = resources.getString(R.string.txtBarcodeNotValid)
            }
            else -> {
                viewModel.openProduct(code, requireActivity())
            }
        }
    }

    @NavigationDrawerType
    override fun getNavigationDrawerType() = NavigationDrawerListener.ITEM_SEARCH_BY_CODE

    override fun onResume() {
        super.onResume()
        (requireActivity() as AppCompatActivity).supportActionBar?.title = getString(R.string.search_by_barcode_drawer)
    }

    override fun onDrawerClosed() {
        super.onDrawerClosed()

        // Force the keyboard to be visible
        if (binding.editTextBarcode.isEmpty()) {
            binding.editTextBarcode.requestFocus()
            requireActivity().showKeyboard()
        }

    }

    companion object {
        const val INTENT_KEY_BARCODE = "barcode"
        fun newInstance(barcode: String? = null) = SearchByCodeFragment().apply {
            arguments = Bundle().apply {
                putString(INTENT_KEY_BARCODE, barcode)
            }
        }
    }
}
